File: proctoring/app/services/proctoring_service.py
"""
Proctoring service for violation detection and handling
"""
from datetime import datetime
from typing import Optional, Tuple
import numpy as np

from app.db.repositories import ExamSessionRepository, ViolationRepository
from app.models.exam import ViolationTypeEnum, ViolationSeverityEnum
from app.proctoring.face_detection import FaceDetector
from app.proctoring.eye_tracking import EyeTracker
from app.proctoring.head_pose import HeadPoseEstimator
from app.proctoring.rules import ViolationRulesEngine
from app.utils.image import base64_to_image
from app.core.logging import logger


class ProctoringService:
    """Service for AI-based exam proctoring"""

    def __init__(self):
        """Initialize proctoring components"""
        self.face_detector = FaceDetector()
        self.eye_tracker = EyeTracker()
        self.head_pose_estimator = HeadPoseEstimator()
        self.rules_engine = ViolationRulesEngine()

    async def verify_initial_identity(
        self,
        session_id: str,
        frame_base64: str,
    ) -> Tuple[bool, Optional[str]]:
        """
        Verify student identity at exam start.
        
        Args:
            session_id: Exam session ID
            frame_base64: Base64 encoded webcam frame
            
        Returns:
            Tuple of (identity_verified, error_message)
        """
        image = base64_to_image(frame_base64)
        if image is None:
            return False, "Failed to decode frame"

        faces, face_count = self.face_detector.detect_faces(image)

        if face_count == 0:
            return False, "No face detected"

        if face_count > 1:
            return False, "Multiple faces detected"

        # Store identity snapshot
        await ExamSessionRepository.verify_identity(session_id, frame_base64)

        logger.info(
            f"Identity verified for session {session_id}",
            extra={"face_count": face_count},
        )

        return True, None

    async def process_proctoring_frame(
        self,
        session_id: str,
        frame_base64: str,
    ) -> Tuple[bool, Optional[str]]:
        """
        Process a frame from the webcam and check for violations.
        
        Args:
            session_id: Exam session ID
            frame_base64: Base64 encoded frame
            
        Returns:
            Tuple of (should_auto_submit, violation_message)
        """
        image = base64_to_image(frame_base64)
        if image is None:
            logger.warning(f"Failed to decode frame for session {session_id}")
            return False, None

        current_time = datetime.utcnow()
        violations_detected = []

        # Face detection check
        faces, face_count = self.face_detector.detect_faces(image)

        if face_count == 0:
            violation = self.rules_engine.check_violation(
                ViolationTypeEnum.NO_FACE,
                condition=True,
                current_time=current_time,
            )
            if violation:
                violations_detected.append(violation)
        elif face_count > 1:
            violation = self.rules_engine.check_violation(
                ViolationTypeEnum.MULTIPLE_FACES,
                condition=True,
                current_time=current_time,
            )
            if violation:
                violations_detected.append(violation)
        else:
            # Single face detected, clear no-face violations
            self.rules_engine.check_violation(
                ViolationTypeEnum.NO_FACE,
                condition=False,
                current_time=current_time,
            )
            self.rules_engine.check_violation(
                ViolationTypeEnum.MULTIPLE_FACES,
                condition=False,
                current_time=current_time,
            )

            # Check eye tracking
            looking_away, gaze_data = self.eye_tracker.detect_eyes_looking_away(image)
            violation = self.rules_engine.check_violation(
                ViolationTypeEnum.LOOKING_AWAY,
                condition=looking_away,
                current_time=current_time,
            )
            if violation:
                violations_detected.append(violation)

            # Check head pose
            head_turned, pose_data = self.head_pose_estimator.estimate_head_pose(image)
            violation = self.rules_engine.check_violation(
                ViolationTypeEnum.HEAD_TURN,
                condition=head_turned,
                current_time=current_time,
            )
            if violation:
                violations_detected.append(violation)

        # Persist violations
        session = await ExamSessionRepository.get_session(session_id)
        if session:
            for violation_data in violations_detected:
                await ViolationRepository.create_violation(
                    session_id=session_id,
                    student_id=session.student_id,
                    violation_type=violation_data["violation_type"],
                    severity=violation_data["severity"],
                    duration_seconds=violation_data.get("duration_seconds"),
                    metadata={
                        "gaze_data": gaze_data if 'gaze_data' in locals() else None,
                        "pose_data": pose_data if 'pose_data' in locals() else None,
                    },
                )

                # Increment warning count
                await ExamSessionRepository.increment_warnings(session_id)

            # Check if should auto-submit
            if self.rules_engine.should_auto_submit():
                return True, f"Auto-submit: {self.rules_engine.warning_count} warnings exceeded"

        return False, None

    def close(self):
        """Close all detector resources"""
        self.face_detector.close()
        self.eye_tracker.close()
        self.head_pose_estimator.close()
        logger.info("Proctoring service closed")

File: proctoring/src/services/proctoringService.js
/**
 * Proctoring service for violation detection and handling
 */
const { v4: uuidv4 } = require('uuid');
const { ExamSession, Violation, ViolationTypeEnum } = require('../models/exam');
const { detectFaces } = require('../proctoring/faceDetection');
const { detectEyesLookingAway } = require('../proctoring/eyeTracking');
const { estimateHeadPose } = require('../proctoring/headPose');
const { createRulesEngine, checkViolation, shouldAutoSubmit } = require('../proctoring/rules');
const { base64ToImage } = require('../utils/image');
const examService = require('./examService');
const logger = require('../config/logger');

// Store rules engines per session
const sessionEngines = new Map();

/**
 * Get or create rules engine for a session
 */
function getEngine(sessionId) {
    if (!sessionEngines.has(sessionId)) {
        sessionEngines.set(sessionId, createRulesEngine());
        logger.info(`Created new rules engine for session: ${sessionId}`);
    }
    return sessionEngines.get(sessionId);
}

/**
 * Remove rules engine for a session
 */
function removeEngine(sessionId) {
    if (sessionEngines.has(sessionId)) {
        sessionEngines.delete(sessionId);
        logger.info(`Removed rules engine for session: ${sessionId}`);
    }
}

/**
 * Verify student identity at exam start
 */
async function verifyInitialIdentity(sessionId, frameBase64) {
    const image = base64ToImage(frameBase64);
    if (image === null) {
        return { verified: false, error: 'Failed to decode frame' };
    }

    // Use async face detection
    const { faces, faceCount, cameraHidden } = await detectFaces(image);

    if (cameraHidden) {
        return { verified: false, error: 'Camera appears to be hidden or blocked' };
    }

    if (faceCount === 0) {
        return { verified: false, error: 'No face detected' };
    }

    if (faceCount > 1) {
        return { verified: false, error: 'Multiple faces detected' };
    }

    // Store identity snapshot
    await examService.verifyIdentity(sessionId, frameBase64);

    logger.info(`Identity verified for session ${sessionId}`, { face_count: faceCount });

    return { verified: true, error: null };
}

/**
 * Process a frame from the webcam and check for violations
 */
async function processProctoringFrame(sessionId, frameBase64) {
    const image = base64ToImage(frameBase64);
    if (image === null) {
        logger.warning(`Failed to decode frame for session ${sessionId}`);
        return { shouldAutoSubmit: false, violationMessage: null };
    }

    const engine = getEngine(sessionId);
    const currentTime = new Date();
    const violationsDetected = [];
    let gazeData = null;
    let poseData = null;
    let landmarks = null;

    // Face detection check (now async)
    const faceResult = await detectFaces(image);
    const { faces, faceCount, cameraHidden } = faceResult;
    landmarks = faceResult.landmarks;
    
    logger.info(`Session ${sessionId}: Face count = ${faceCount}, cameraHidden = ${cameraHidden}`);

    // Check for camera hidden/blocked
    if (cameraHidden) {
        logger.warning(`Session ${sessionId}: Camera appears to be hidden/blocked`);
        const violation = checkViolation(engine, ViolationTypeEnum.NO_FACE, true, currentTime);
        if (violation) {
            violation.metadata = { reason: 'camera_hidden' };
            violationsDetected.push(violation);
        }
    } else if (faceCount === 0) {
        logger.info(`Session ${sessionId}: No face detected, checking violation rules...`);
        const violation = checkViolation(engine, ViolationTypeEnum.NO_FACE, true, currentTime);
        if (violation) {
            logger.warning(`Session ${sessionId}: NO_FACE violation triggered!`);
            violationsDetected.push(violation);
        }
    } else if (faceCount > 1) {
        const violation = checkViolation(engine, ViolationTypeEnum.MULTIPLE_FACES, true, currentTime);
        if (violation) {
            violationsDetected.push(violation);
        }
    } else {
        // Single face detected, clear no-face violations
        checkViolation(engine, ViolationTypeEnum.NO_FACE, false, currentTime);
        checkViolation(engine, ViolationTypeEnum.MULTIPLE_FACES, false, currentTime);

        // Get landmarks from first face
        const faceLandmarks = landmarks && landmarks.length > 0 ? landmarks[0] : null;

        // Check eye tracking with landmarks
        const eyeResult = detectEyesLookingAway(image, faceLandmarks);
        gazeData = eyeResult.gazeData;
        const violation1 = checkViolation(engine, ViolationTypeEnum.LOOKING_AWAY, eyeResult.lookingAway, currentTime);
        if (violation1) {
            violationsDetected.push(violation1);
        }

        // Check head pose with landmarks
        const poseResult = estimateHeadPose(image, faceLandmarks);
        poseData = poseResult.poseData;
        const violation2 = checkViolation(engine, ViolationTypeEnum.HEAD_TURN, poseResult.headTurned, currentTime);
        if (violation2) {
            violationsDetected.push(violation2);
        }
    }

    // Persist violations
    const session = await ExamSession.findById(sessionId);
    const violationMessages = [];

    if (session) {
        for (const violationData of violationsDetected) {
            const violationId = uuidv4();
            const violation = new Violation({
                _id: violationId,
                session_id: sessionId,
                student_id: session.student_id,
                violation_type: violationData.violation_type,
                severity: violationData.severity,
                timestamp: new Date(),
                duration_seconds: violationData.duration_seconds,
                metadata: {
                    gaze_data: gazeData,
                    pose_data: poseData,
                    ...violationData.metadata,
                },
            });

            await violation.save();
            logger.info(`Violation created: ${violationId}`);

            // Increment warning count
            await examService.incrementWarnings(sessionId);

            // Build violation message for frontend
            violationMessages.push(`${violationData.violation_type} detected`);
        }

        // Check if should auto-submit
        if (shouldAutoSubmit(engine)) {
            return {
                shouldAutoSubmit: true,
                violationMessage: `Auto-submit: ${engine.warningCount} warnings exceeded`,
            };
        }
    }

    // Return violation message if any violations were detected
    if (violationMessages.length > 0) {
        return {
            shouldAutoSubmit: false,
            violationMessage: violationMessages.join('; '),
        };
    }

    return { shouldAutoSubmit: false, violationMessage: null };
}

module.exports = {
    getEngine,
    removeEngine,
    verifyInitialIdentity,
    processProctoringFrame,
};
